### DISCUSSION TITLE
Getting the boundary condition for `∂z(b)` in a doubly-periodic LES?

### DISCUSSION THREAD

**Discussion by tomchor** [CliMA/Oceananigans.jl#2950] [General]
I have a LES set-up kinda similar to the [ocean mixing example](https://clima.github.io/OceananigansDocumentation/stable/generated/ocean_wind_mixing_and_convection/), and the important part is that I'm setting a flux as the top BC for buoyancy. One of the things I'm trying to calculate is the vertical buoyancy gradient, but with boundary conditions that are as close to correct as possible. So basically I'm trying to get:

```julia
@at (Center, Center, Center) ∂z(b)
```
with "proper" boundary conditions. (It's important that it's on cell centers.)

However, I haven't been able to do that properly. The first question that comes in mind is: Is it possible to have an appropriate BC for db/dz when the only closure is an LES? (By that, I mean without any constant background viscosity/diffusivity.) I can't think of a way to do it without estimating the eddy diffusivity at the boundary, where it isn't defined, so I'd say that the answer is no. However, I'm interested to see if anyone has ideas or anything to add here that I may be missing.

With that said, I believe a physically-correct way to do it would *need* a constant eddy viscosity in the background (molecular or not). Then we could use only that viscosity at the boundary to get the gradient (by inverting the flux-gradient relationship since we have the flux at the boundary). 

But even if I take this approach, it doesn't appear to have any effect on the result. For example, in the MWE below I'm setting up a model with a surface buoyancy flux and then I'm calculating `∂z(b)` in two ways:

- `dbdz1` just naively calculates `∂z(b)` without any boundary conditions
- `dbdz2` tries to impose the BCs in the way I described above.

Here's what this looks like for me (sorry if it's a bit convoluted):

```julia
using Oceananigans 
using Oceananigans.AbstractOperations: @at, ∂z
using Oceananigans.Fields: @compute

grid = RectilinearGrid(size = (8, 8, 16), extent=(400, 400, 100))

b_bc = FieldBoundaryConditions(top = FluxBoundaryCondition(3e-9))
model = NonhydrostaticModel(; grid, 
                            tracers=:b, 
                            buoyancy=BuoyancyTracer(),
                            closure=(ScalarDiffusivity(ν=1e-5, κ=1e-5), AnisotropicMinimumDissipation()),
                            boundary_conditions = (; b=b_bc))

constant_stratification(x, y, z) = 1e-5 * z
noise(x, y, z) = 1e-2 * randn()
set!(model, b=constant_stratification, w=noise)

u, v, w = model.velocities
b = model.tracers.b
    
@compute dbdz1 = Field(@at (Center, Center, Center) ∂z(b))

bz_bcs = FieldBoundaryConditions(grid, (Center, Center, Center);
                                 top = ValueBoundaryCondition(-model.tracers.b.boundary_conditions.top.condition / model.closure[1].κ.b))
@compute dbdz2 = Field((@at (Center, Center, Center) ∂z(b)), boundary_conditions=bz_bcs)
```

However, it appears the boundary condition doesn't take any effect. After running this MWE I get:

```julia
julia> dbdz1 == dbdz2
true

julia> interior(dbdz2)[:,:,end]
8×8 Matrix{Float64}:
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
 5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6  5.0e-6
```

Note that `5.0e-6` is exactly half of the the initial stratification, indicating that this is just a linear interpolation between the interior stratification and zero.

Am I missing something in setting the BC?

**Comment by glwagner** [comment]
I think the issue here is that you really would like to impose boundary conditions on `bz = Field(∂z(b))` with location `(Center, Center, Face)`. 

You can try that:

```julia
bz_bcs = FieldBoundaryConditions(grid, (Center, Center, Face);
                                 top = OpenBoundaryCondition(-model.tracers.b.boundary_conditions.top.condition / model.closure[1].κ.b))
bz = Field(∂z(b)), boundary_conditions=bz_bcs)

bz_ccc_op = @at (Center, Center, Center) bz # I _think_ this will work, but you may have to define this interpolation manually
bz_ccc = Field(bz_ccc_op)
```

Probably a more efficient solution is something like

```julia
κ = 1e-5
bz_top = 3e-9 / κ # or whatever this needs to be
# create model with `FluxBoundaryCondition` on `b` as above

# Create an "auxiliary" buoyancy field with `GradientBoundaryCondition`:
bz_top_bc = GradientBoundaryCondition(bz_top)
auxiliary_b_bcs = FieldBoundaryConditions(grid, (Center, Center, Center); top = bz_top_bc)
b_gradient_bc = CenterField(grid; data=model.tracers.b, boundary_conditions=auxiliary_b_bcs)

# The desired field:
dbdz = Field(@at (Center, Center, Center) ∂z(b_gradient_bcs))
```

The above approach may only work if `fill_halo_regions!(b_gradient_bc)` is called prior to calling `compute!(dbdz)`.

I think though what is really desired here is to set the boundary conditions on the AMD effective diffusivity to `ValueBoundaryCondition(0)`. In that case, you can use `GradientBoundaryCondition` directly on buoyancy via the model constructor.
This is actually part of the code, though it is not tested (because I wasn't sure it would ever be used). However it looks like it might work because:

https://github.com/CliMA/Oceananigans.jl/blob/4923669bc00337191542cbb882e1347457c16a2f/src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl#L347-L359

so you can try adding the boundary condition

```julia
κₑ_b_bcs = FieldBoundaryConditions(grid, (Center, Center, Center), top=ValueBoundaryCondition(0)) 
boundary_conditions = (b=b_bcs, κₑ=(; b=κₑ_b_bcs))
```

Not the easiest to use interface... perhaps we can brainstorm on how to make this a little easier for users.

**Reply by tomchor** [reply]
> I think the issue here is that you really would like to impose boundary conditions on `bz = Field(∂z(b))` with location `(Center, Center, Face)`.
> 
> You can try that:
> 
> ```julia
> bz_bcs = FieldBoundaryConditions(grid, (Center, Center, Face);
>                                  top = ValueBoundaryCondition(-model.tracers.b.boundary_conditions.top.condition / model.closure[1].κ.b))
> bz = Field(∂z(b)), boundary_conditions=bz_bcs)
> 
> bz_ccc_op = @at (Center, Center, Center) bz # I _think_ this will work, but you may have to define this interpolation manually
> bz_ccc = Field(bz_ccc_op)
> ```

This worked! Thanks. I'll probably stick to this solution since it's simpler for me to understand and it's not contingent to filling halos.

I understand why we'd want to apply the BC on `Center, Center, Center`, but I gotta say, I still don't understand why the BC gets ignored in the example I posted. Is this expected? Since `Field()` is receiving an `AbstractOperation` that results in a `Center, Center, Center` field, I don't see why it doesn't apply a BC that's defined as `FieldBoundaryConditions(grid, (Center, Center, Center); kwargs...)`.

**Reply by tomchor** [reply]
> I think though what is really desired here is to set the boundary conditions on the AMD effective diffusivity to `ValueBoundaryCondition(0)`. In that case, you can use `GradientBoundaryCondition` directly on buoyancy via the model constructor. This is actually part of the code, though it is not tested (because I wasn't sure it would ever be used). However it looks like it might work because:
> 
> https://github.com/CliMA/Oceananigans.jl/blob/4923669bc00337191542cbb882e1347457c16a2f/src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl#L347-L359
> 
> so you can try adding the boundary condition
> 
> ```julia
> κₑ_b_bcs = FieldBoundaryConditions(grid, (Center, Center, Center), top=ValueBoundaryCondition(0)) 
> boundary_conditions = (b=b_bcs, κₑ=(; b=κₑ_b_bcs))
> ```
> 
> Not the easiest to use interface... perhaps we can brainstorm on how to make this a little easier for users.

But including this BC for `κₑ` would make it zero on the center node closest to the boundary, which isn't correct no? `κₑ` is only zero _at_ the boundary (meaning in this case a `Center, Center, Face` location).

**Reply by glwagner** [reply]
Hmm no. Value and Gradient boundary conditions only applied to fields at Center locations. `ValueBoundaryCondition(c)` means that a field is reconstructed / interpolated to `c` on the boundary.

I updated my answer above which incorrectly applied `ValueBoundaryCondition` to a Face field.

**Reply by tomchor** [reply]
> Hmm no. Value and Gradient boundary conditions only applied to fields at Center locations. `ValueBoundaryCondition(c)` means that a field is reconstructed / interpolated to `c` on the boundary.

Ah, okay. I guess I was misunderstanding how value BCs were applied to center fields this whole time! But yeah, it makes sense that passing a BC to $\kappa_e$ is probably the best way to do this. In fact, why don't we set those BCs by default for Smag and AMD? I think physically those values _should_ be zero at the boundaries. Can't we make that the default (only for AMD and Smag that is)?

**Reply by glwagner** [reply]
Most of the papers I read that mentioned boundary conditions use a Neumann / no gradient boundary condition at the wall for eddy diffusivities. Maybe we can make it an option of the closure to make things a little easier.

**Reply by glwagner** [reply]
For the majority of ocean LES applications at realistic Reynolds numbers I think best practice is to omit molecular diffusion entirely, because it's negligible. It's better practice because it makes setups / science more reproducible: we don't need to bandy around parameters / constants that are irrelevant to the outcome of a simulation. In the same vein, realistic ocean LES are almost wall-modeled rather than wall-resolved and therefore the eddy diffusivities should not be zero at the "wall". So I don't agree that we should make this the default.

However there are obviously situations (like wall-resolved LES) in which one needs molecular diffusion, and in which one may want to pin the diffusivity to its molecular value at the boundary. So we should support that.

We could even have the user pass boundary conditions into the closure? Trying to think what would yield the most interpretable script / API.
